include/llama.h

```diff
@@ -1007,158 +997,136 @@ extern "C" {
+    //
+    //   - RNG state (seed and generator)
+    //   - Custom set of constraints (see llama_sampler_constraint_add)
+    //   - Sampling method (greedy, dist)
+    //   - Previous tokens
+    //
+    // In the future, it will be utilized offload the sampling to the backends (e.g. GPU).
+    //
+    // TODO: in the future, the entire API should be changed to accept llama_vocab, instead of llama_model
 
-    /// Initialize a llama_grammar.
-    ///
-    /// @param rules The rule elements of the grammar to initialize.
-    /// @param n_rules The number of rules.
-    /// @param start_rule_index The index of the root rule (the starting point of the grammar).
-    /// @return The initialized llama_grammar or nullptr if initialization failed.
-    LLAMA_API struct llama_grammar * llama_grammar_init(
-            const llama_grammar_element ** rules,
-                                 size_t    n_rules,
-                                 size_t    start_rule_index);
-
-    LLAMA_API void llama_grammar_free(struct llama_grammar * grammar);
-
-    LLAMA_API struct llama_grammar * llama_grammar_copy(const struct llama_grammar * grammar);
-
-    /// @details Apply constraints from grammar
-    LLAMA_API void llama_grammar_sample(
-            const struct llama_grammar * grammar,
-            const struct llama_context * ctx,
-                llama_token_data_array * candidates);
-    LLAMA_API DEPRECATED(void llama_sample_grammar(
-            struct llama_context * ctx,
-          llama_token_data_array * candidates,
-      const struct llama_grammar * grammar),
-        "use llama_grammar_sample instead");
+    // constraints
 
-    /// @details Accepts the sampled token into the grammar
-    LLAMA_API void llama_grammar_accept_token(
-            struct llama_grammar * grammar,
-            struct llama_context * ctx,
-                     llama_token   token);
+    struct llama_constraint;
 
-    //
-    // Sampling functions
-    //
+    typedef void * llama_constraint_context_t;
 
-    // Sets the current rng seed.
-    LLAMA_API void llama_set_rng_seed(struct llama_context * ctx, uint32_t seed);
+    // user code can implement the interface below in order to create custom llama_constraint
+    struct llama_constraint_i {
+        const char *              (*name)  (const struct llama_constraint * cnstr);                                 // can be NULL
+        void                      (*accept)(      struct llama_constraint * cnstr, llama_token token);              // can be NULL
+        void                      (*apply) (      struct llama_constraint * cnstr, llama_token_data_array * cur_p); // required
+        void                      (*reset) (      struct llama_constraint * cnstr);                                 // can be NULL
+        struct llama_constraint * (*copy)  (const struct llama_constraint * cnstr);                                 // can be NULL if ctx is NULL
```

Comment:
A more common name for this functionality is clone. This applies to the other functions, `llama_constraint_clone` and `llama_sampler_clone` may be more intuitive.